content/blog/2019-09-25-interpreter-vector-support/index.md

```diff
@@ -0,0 +1,134 @@
++++
+title = "Vector Instruction Support in the Bril Interpreter"
+extra.bio = """
+  Philip Bedoukian is a 3rd year PhD student in ECE. His research focuses on reconfigurable hardware.
+"""
+extra.author = "Philip Bedoukian"
++++
+
+## Problem Statement
+
+The Bril interpreter does not take advantage of the vector instructions available in modern CPUs. This presents two problems: the first is that a Bril backend cannot generate vector instructions because they are not present in the intermediate language. The second issue is that  interpreting vector instructions is slow because it requires a loop over each vector element. We propose to support vector instructions in the interpreter and accelerate the interpretation of them using intrinsic vector instructions.
+
+## Vector Instructions
+
+Single-Instruction Multiple-Data (SIMD) allows computer hardware to obtain signifigant speedups over the conventional execution paradigm, Multiple-Instruction Multiple-Data (MIMD). SIMD executes fewer instructions than MIMD to accomplish the same amount of work. In SIMD, multiple arithmetic operations are grouped together under one instruction. Generally, this aritmetic is executed over multiple execution units at the same instant (spatial SIMD), although some implementations allow the same instruction to use the same functional unit over multiple cycles (temporal SIMD). Various hardware architectures allow the programmer to support SIMD by exposing vector instructions in their ISA. Typically, there is a vector register file which holds multiple elements per register and can participate in vector arithmetic.
+
+## Vector Support in the Bril Interpreter
+
+### Memory Support
+
+Vector instructions are only useful when operating on large chunks of data. Conversely, registers are not designed to hold this amount of data. A prerequisite for vector instructions is support for data memory in the interpreter. The interpreter memory is implemented like a stack. The program is allowed to access any location in the fixed size memory. Stack frames were not implemented because the interpreter only supports a single function.
+
+Memory access requires both loads and store operations. To emulate arrays, we can loop over multiple addresses and perform a load and/or store at each location. Load and store instructions in Bril are implemented similarly to their assembly counterparts. A load takes in a memory address (an `int` register in Bril) and writes to a destination register. A store is an effect operation and does not write to a register. It uses two source registers: a register containing a value and a register containing an address. The Bril syntax is given below.
```

Comment:
```suggestion
Memory access requires both load and store operations. To emulate arrays, we can loop over multiple addresses and perform a load and/or store at each location. Load and store instructions in Bril are implemented similarly to their assembly counterparts. A load takes in a memory address (an `int` register in Bril) and writes to a destination register. A store is an effect operation and does not write to a register. It uses two source registers: a register containing a value and a register containing an address. The Bril syntax is given below.
```

Reply:
Maybe it's worth mentioning the names of the instructions, `lw` and `sw`?

Also, FWIW, you might be interested in this other design with arrays, loads, and stores:
https://www.cs.cornell.edu/courses/cs6120/2019fa/blog/manually-managed-memory/